<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>linkSet • linkSet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="linkSet">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">linkSet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/linkSet.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/hic_workthrough.html">hic_workthrough</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GilbertHan1011/linkSet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>linkSet</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/GilbertHan1011/linkSet/blob/HEAD/vignettes/linkSet.Rmd" class="external-link"><code>vignettes/linkSet.Rmd</code></a></small>
      <div class="d-none name"><code>linkSet.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>With techniche like HiC, PCHiC, scATAC-seq, and other high-throughput
omics techniques, we are able to predict the interactions between genes
and enhancers. The linkSet package provides classes to represent the
interactions between genes and enhancers, and provides functions to
analyze and visualize the interactions.</p>
</div>
<div class="section level2">
<h2 id="the-linkset-class">The linkSet class<a class="anchor" aria-label="anchor" href="#the-linkset-class"></a>
</h2>
<p>The linkSet class is a specialized data structure designed to
represent and analyze genomic interactions, particularly focusing on
gene-enhancer relationships. It’s part of the linkSet package, which
provides tools for managing and analyzing genomic link data in
bioinformatics and genomic research.</p>
<p>Key features of the linkSet class:</p>
<ol style="list-style-type: decimal">
<li>Representation of genomic interactions:</li>
</ol>
<ul>
<li>It stores two types of genomic anchors: “bait” (typically genes) and
“other end” (oe, typically enhancers or other regulatory elements).</li>
<li>These anchors are represented as genomic ranges, allowing for
precise localization on chromosomes.</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>Flexible input: Can be constructed from various data types,
including GRanges objects for both anchors, or character vectors for
bait and GRanges for other ends. Supports conversion from other common
genomic data structures like GInteractions and data frames.</li>
<li>Metadata storage: Allows for additional information to be stored
alongside the genomic interactions, such as interaction scores or
experimental conditions.</li>
<li>Biological context: Designed to work with data from various
high-throughput genomic techniques like Hi-C, Promoter Capture Hi-C
(PCHi-C), and scATAC-seq. Facilitates the analysis of long-range
chromatin interactions, which are crucial for understanding gene
regulation and 3D genome organization.</li>
<li>Annotation capabilities: Includes methods for annotating promoters
and distinguishing between inter- and intra-chromosomal interactions.
This feature is particularly useful for identifying potential regulatory
relationships between distal genomic elements.</li>
<li>Integration with Bioconductor: Built on top of established
Bioconductor classes, ensuring compatibility with a wide range of
genomic analysis tools.</li>
<li>Visualization and analysis: The class is designed to support
functions for analyzing and visualizing genomic interactions.</li>
</ol>
<p><img src="img/overview.png" alt="Overview diagram of the linkSet class structure and its components" width="70%"></p>
</div>
<div class="section level2">
<h2 id="construction">Construction<a class="anchor" aria-label="anchor" href="#construction"></a>
</h2>
<p>The linkSet class can be constructed from various data types,
including GRanges objects for both anchors, or character vectors for
bait and GRanges for other ends.
<img src="img/9.10_data_structure.png" alt="data structure of the linkSet class" width="70%"></p>
<div class="section level3">
<h3 id="construction-from-granges-objects">Construction from GRanges objects<a class="anchor" aria-label="anchor" href="#construction-from-granges-objects"></a>
</h3>
<p>To construct a <code>linkSet</code> object from GRanges objects:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GilbertHan1011/linkSet" class="external-link">linkSet</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="va">gr1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>              ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">100</span>, <span class="fl">200</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>              strand <span class="op">=</span> <span class="st">"+"</span>, symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gr2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>              ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">150</span>, <span class="fl">250</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>              strand <span class="op">=</span> <span class="st">"+"</span><span class="op">)</span></span>
<span><span class="co">## construct linkSet</span></span>
<span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/linkSet/man/linkSet-class.html" class="external-link">linkSet</a></span><span class="op">(</span><span class="va">gr1</span>, <span class="va">gr2</span>, specificCol <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="va">ls</span></span>
<span><span class="co">#&gt; linkSet object with 3 interactions and 1 metadata column:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]       Gene1 ---        chr1     50-59 |          Gene1</span></span>
<span><span class="co">#&gt;   [2]       Gene2 ---        chr2   150-159 |          Gene2</span></span>
<span><span class="co">#&gt;   [3]       Gene3 ---        chr2   250-259 |          Gene3</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 6 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="construction-from-ginteractions">Construction from GInteractions<a class="anchor" aria-label="anchor" href="#construction-from-ginteractions"></a>
</h3>
<p>You can also create a <code>linkSet</code> object from a
<code>GInteractions</code> object using the <code>Convert</code>
function:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'InteractionSet'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:linkSet':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anchorIds, anchors, pairdist, reduceRegions, regions, regions&lt;-</span></span>
<span><span class="va">gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/GInteractions-class.html" class="external-link">GInteractions</a></span><span class="op">(</span></span>
<span>  anchor1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">5</span><span class="op">)</span>,</span>
<span>  anchor2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>  regions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span>, <span class="st">"chr3"</span><span class="op">)</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">50</span>, <span class="fl">100</span>, <span class="fl">150</span>, <span class="fl">200</span>, <span class="fl">250</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sox2"</span>, <span class="st">"Sox9"</span>, <span class="st">"Sox10"</span><span class="op">)</span></span>
<span><span class="co">##Convert to linkSet</span></span>
<span><span class="va">ls_from_gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Convert.html">Convert</a></span><span class="op">(</span><span class="va">gi</span>, baitCol <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="va">ls_from_gi</span></span>
<span><span class="co">#&gt; linkSet object with 3 interactions and 2 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe |     score        gene</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; | &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]        Sox2 ---        chr1     50-59 |       0.1        Sox2</span></span>
<span><span class="co">#&gt;   [2]        Sox9 ---        chr2   150-159 |       0.2        Sox9</span></span>
<span><span class="co">#&gt;   [3]       Sox10 ---        chr3   250-259 |       0.3       Sox10</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 6 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>Some times, you need to use gene and enhancer information to
construct a <code>linkSet</code> object from GInteractions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneGr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span><span class="op">)</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">105</span>, <span class="fl">205</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,</span>
<span>  symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">peakGr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span><span class="op">)</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">45</span>, <span class="fl">145</span>, <span class="fl">245</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run baitGInteractions</span></span>
<span><span class="va">ls_from_gi_2</span> <span class="op">&lt;-</span> <span class="fu">baitGInteractions</span><span class="op">(</span><span class="va">gi</span>, <span class="va">geneGr</span>, <span class="va">peakGr</span>, geneSymbol <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ls_from_gi_2</span></span>
<span><span class="co">#&gt; linkSet object with 3 interactions and 3 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol     score</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;   [1]       Gene1 ---        chr1     45-54 |          Gene1       0.1</span></span>
<span><span class="co">#&gt;   [2]       Gene2 ---        chr2   145-154 |          Gene2       0.2</span></span>
<span><span class="co">#&gt;   [3]       Gene3 ---        chr3   245-254 |          Gene3       0.3</span></span>
<span><span class="co">#&gt;              gene</span></span>
<span><span class="co">#&gt;       &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]        Sox2</span></span>
<span><span class="co">#&gt;   [2]        Sox9</span></span>
<span><span class="co">#&gt;   [3]       Sox10</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 6 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="other-construction-methods">Other construction methods<a class="anchor" aria-label="anchor" href="#other-construction-methods"></a>
</h3>
<p>You can also construct a <code>linkSet</code> object from a data
frame.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span>,</span>
<span>  peak <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:1000-2000"</span>, <span class="st">"chr2:3000-4000"</span>, <span class="st">"chr3:5000-6000"</span><span class="op">)</span>,</span>
<span>  score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.7</span>, <span class="fl">0.9</span><span class="op">)</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Test successful conversion</span></span>
<span><span class="va">ls_from_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Convert.html">Convert</a></span><span class="op">(</span><span class="va">test_df</span><span class="op">)</span></span>
<span><span class="va">ls_from_df</span></span>
<span><span class="co">#&gt; linkSet object with 3 interactions and 1 metadata column:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe |     score</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;   [1]       gene1 ---        chr1  999-2000 |       0.5</span></span>
<span><span class="co">#&gt;   [2]       gene2 ---        chr2 2999-4000 |       0.7</span></span>
<span><span class="co">#&gt;   [3]       gene3 ---        chr3 4999-6000 |       0.9</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 3 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="accessors">Accessors<a class="anchor" aria-label="anchor" href="#accessors"></a>
</h2>
<div class="section level3">
<h3 id="getters">Getters<a class="anchor" aria-label="anchor" href="#getters"></a>
</h3>
<p>The important elements of the <code>linkSet</code> object can be
easily accessed by the following accessors:</p>
<ul>
<li>
<code>regions(x)</code>: Get the regions of the <code>linkSet</code>
object.</li>
<li>
<code>bait(x)</code>: Get the bait of the <code>linkSet</code>
object.</li>
<li>
<code>oe(x)</code>: Get the other end of the <code>linkSet</code>
object.</li>
<li>
<code>regionsBait(x)</code>: Get the bait regions of the
<code>linkSet</code> object.</li>
<li>
<code>regionsOe(x)</code>: Get the other end regions of the
<code>linkSet</code> object.</li>
<li>
<code>mcols(x)</code>: Get the metadata of the <code>linkSet</code>
object.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">linkSet</span><span class="fu">::</span><span class="fu"><a href="../reference/linkSet-accessors.html">regions</a></span><span class="op">(</span><span class="va">ls</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 6 ranges and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames    ranges strand</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr1      1-10      +</span></span>
<span><span class="co">#&gt;   [2]     chr1     50-59      +</span></span>
<span><span class="co">#&gt;   [3]     chr1   100-109      +</span></span>
<span><span class="co">#&gt;   [4]     chr2   150-159      +</span></span>
<span><span class="co">#&gt;   [5]     chr2   200-209      +</span></span>
<span><span class="co">#&gt;   [6]     chr2   250-259      +</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">regionsBait</span><span class="op">(</span><span class="va">ls</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 3 ranges and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames    ranges strand</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr1      1-10      +</span></span>
<span><span class="co">#&gt;   [2]     chr1   100-109      +</span></span>
<span><span class="co">#&gt;   [3]     chr2   200-209      +</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">oe</span><span class="op">(</span><span class="va">ls</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 3 ranges and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames    ranges strand</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr1     50-59      +</span></span>
<span><span class="co">#&gt;   [2]     chr2   150-159      +</span></span>
<span><span class="co">#&gt;   [3]     chr2   250-259      +</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">bait</span><span class="op">(</span><span class="va">ls</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Gene1" "Gene2" "Gene3"</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">ls</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 3 rows and 1 column</span></span>
<span><span class="co">#&gt;   anchor1.symbol</span></span>
<span><span class="co">#&gt;      &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1          Gene1</span></span>
<span><span class="co">#&gt; 2          Gene2</span></span>
<span><span class="co">#&gt; 3          Gene3</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="setters">Setters<a class="anchor" aria-label="anchor" href="#setters"></a>
</h3>
<p>Wait… You set wrong bait or oe region? You can also change it
easily:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_bait</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene40"</span>,<span class="st">"Gene41"</span>,<span class="st">"Gene42"</span><span class="op">)</span></span>
<span><span class="va">new_oe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span><span class="op">)</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">105</span>, <span class="fl">205</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,</span>
<span>  symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">bait</span><span class="op">(</span><span class="va">ls</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">new_bait</span></span>
<span><span class="fu">oe</span><span class="op">(</span><span class="va">ls</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">new_oe</span></span>
<span><span class="va">ls</span></span>
<span><span class="co">#&gt; linkSet object with 3 interactions and 2 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol      symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;   [2]      Gene41 ---        chr2   105-124 |          Gene2       Gene2</span></span>
<span><span class="co">#&gt;   [3]      Gene42 ---        chr3   205-224 |          Gene3       Gene3</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 6 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="subsetting-and-concatenation">Subsetting and concatenation<a class="anchor" aria-label="anchor" href="#subsetting-and-concatenation"></a>
</h2>
<p>LinkSet object can be easily subsetted by index.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ls_sub</span> <span class="op">&lt;-</span> <span class="va">ls</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">ls_sub</span></span>
<span><span class="co">#&gt; linkSet object with 2 interactions and 2 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol      symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;   [2]      Gene41 ---        chr2   105-124 |          Gene2       Gene2</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 6 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>If you are interested in a specific gene or region, you can subset
the <code>linkSet</code> object by the bait or oe region.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ls_sub_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/linkSet-subset-methods.html">subsetBait</a></span><span class="op">(</span><span class="va">ls</span>, <span class="st">"Gene1"</span><span class="op">)</span></span>
<span><span class="va">ls_sub_2</span></span>
<span><span class="co">#&gt; linkSet object with 0 interactions and 2 metadata columns:</span></span>
<span><span class="co">#&gt;           bait     seqnames_oe ranges_oe | anchor1.symbol      symbol</span></span>
<span><span class="co">#&gt;    &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 0 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="va">ls_sub_3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/linkSet-subset-methods.html">subsetBaitRegion</a></span><span class="op">(</span><span class="va">ls</span>, <span class="st">"chr1:100-200"</span><span class="op">)</span></span>
<span><span class="va">ls_sub_3</span></span>
<span><span class="co">#&gt; linkSet object with 1 interaction and 2 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol      symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]      Gene41 ---        chr2   105-124 |          Gene2       Gene2</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 2 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>You can also concatenate two <code>linkSet</code> objects.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ls_concat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ls</span>, <span class="va">ls</span><span class="op">)</span></span>
<span><span class="va">ls_concat</span></span>
<span><span class="co">#&gt; linkSet object with 6 interactions and 2 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol      symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;   [2]      Gene41 ---        chr2   105-124 |          Gene2       Gene2</span></span>
<span><span class="co">#&gt;   [3]      Gene42 ---        chr3   205-224 |          Gene3       Gene3</span></span>
<span><span class="co">#&gt;   [4]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;   [5]      Gene41 ---        chr2   105-124 |          Gene2       Gene2</span></span>
<span><span class="co">#&gt;   [6]      Gene42 ---        chr3   205-224 |          Gene3       Gene3</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 6 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="granges-method">GRanges method<a class="anchor" aria-label="anchor" href="#granges-method"></a>
</h2>
<p>The features of linkSet is inherited from GRanges, which means you
can use all the functions in GRanges to manipulate the linkSet object.
The most wonderful things is that you can change the bait region and oe
region separately.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">linkExample</span><span class="op">)</span></span>
<span><span class="va">linkExample</span></span>
<span><span class="co">#&gt; linkSet object with 5 interactions and 1 metadata column:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]       Gene1 ---        chr1     50-59 |          Gene1</span></span>
<span><span class="co">#&gt;   [2]       Gene1 ---        chr2   150-159 |          Gene1</span></span>
<span><span class="co">#&gt;   [3]       Gene2 ---        chr2   250-259 |          Gene2</span></span>
<span><span class="co">#&gt;   [4]       Gene3 ---        chr4   350-359 |          Gene3</span></span>
<span><span class="co">#&gt;   [5]       Gene3 ---        chr4   450-459 |          Gene3</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 10 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 4 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co">## resize the bait region</span></span>
<span><span class="va">resize_bait</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/linkSet-GRange-Methods.html">resizeRegions</a></span><span class="op">(</span><span class="va">linkExample</span>, width <span class="op">=</span> <span class="fl">75</span>, fix <span class="op">=</span> <span class="st">"start"</span>, region <span class="op">=</span> <span class="st">"bait"</span><span class="op">)</span></span>
<span><span class="va">resize_bait</span></span>
<span><span class="co">#&gt; linkSet object with 5 interactions and 1 metadata column:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]       Gene1 ---        chr1     50-59 |          Gene1</span></span>
<span><span class="co">#&gt;   [2]       Gene1 ---        chr2   150-159 |          Gene1</span></span>
<span><span class="co">#&gt;   [3]       Gene2 ---        chr2   250-259 |          Gene2</span></span>
<span><span class="co">#&gt;   [4]       Gene3 ---        chr4   350-359 |          Gene3</span></span>
<span><span class="co">#&gt;   [5]       Gene3 ---        chr4   450-459 |          Gene3</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 10 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 4 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co">## resize the oe region</span></span>
<span><span class="va">resize_oe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/linkSet-GRange-Methods.html">resizeRegions</a></span><span class="op">(</span><span class="va">linkExample</span>, width <span class="op">=</span> <span class="fl">75</span>, fix <span class="op">=</span> <span class="st">"start"</span>, region <span class="op">=</span> <span class="st">"oe"</span><span class="op">)</span></span>
<span><span class="va">resize_oe</span></span>
<span><span class="co">#&gt; linkSet object with 5 interactions and 1 metadata column:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]       Gene1 ---        chr1    50-124 |          Gene1</span></span>
<span><span class="co">#&gt;   [2]       Gene1 ---        chr2   150-224 |          Gene1</span></span>
<span><span class="co">#&gt;   [3]       Gene2 ---        chr2   250-324 |          Gene2</span></span>
<span><span class="co">#&gt;   [4]       Gene3 ---        chr4   350-424 |          Gene3</span></span>
<span><span class="co">#&gt;   [5]       Gene3 ---        chr4   450-524 |          Gene3</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 10 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 4 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>The <code>reduce</code> method allows the <code>linkSet</code>
objects to be collapsed across the whole of the <code>linkSet</code>
object.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reduce_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/linkSet-GRange-Methods.html">reduce</a></span><span class="op">(</span><span class="va">ls_concat</span><span class="op">)</span></span>
<span><span class="va">reduce_ls</span></span>
<span><span class="co">#&gt; linkSet object with 3 interactions and 3 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol      symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;   [2]      Gene41 ---        chr2   105-124 |          Gene2       Gene2</span></span>
<span><span class="co">#&gt;   [3]      Gene42 ---        chr3   205-224 |          Gene3       Gene3</span></span>
<span><span class="co">#&gt;           count</span></span>
<span><span class="co">#&gt;       &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]         2</span></span>
<span><span class="co">#&gt;   [2]         2</span></span>
<span><span class="co">#&gt;   [3]         2</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 6 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>By default, we will collapse the duplicate interactions, and count
the interactions and store in the metadata columns. You can also choose
not to count the interactions by set
<code>countInteractions = FALSE</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reduce_ls_2</span> <span class="op">&lt;-</span> <span class="fu">linkSet</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html" class="external-link">reduceRegions</a></span><span class="op">(</span><span class="va">ls_concat</span>, countInteractions <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">reduce_ls_2</span></span>
<span><span class="co">#&gt; linkSet object with 6 interactions and 2 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol      symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;   [2]      Gene41 ---        chr2   105-124 |          Gene2       Gene2</span></span>
<span><span class="co">#&gt;   [3]      Gene42 ---        chr3   205-224 |          Gene3       Gene3</span></span>
<span><span class="co">#&gt;   [4]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;   [5]      Gene41 ---        chr2   105-124 |          Gene2       Gene2</span></span>
<span><span class="co">#&gt;   [6]      Gene42 ---        chr3   205-224 |          Gene3       Gene3</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 6 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="diagnose">Diagnose<a class="anchor" aria-label="anchor" href="#diagnose"></a>
</h2>
<p>There are two metrics to diagnose the quality of the
<code>linkSet</code> object: 1. The intra/inter-chromosomal
interactions. 2. The distance between the bait and oe region.</p>
<p>The <code>diagnose</code> function can help you to diagnose the
<code>linkSet</code> object.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">diagnoseLinkSet</span><span class="op">(</span><span class="va">linkExample</span><span class="op">)</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre></div>
<p><img src="linkSet_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; linkSet object with 5 interactions and 3 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol  inter_type</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]       Gene1 ---        chr1     50-59 |          Gene1       inter</span></span>
<span><span class="co">#&gt;   [2]       Gene1 ---        chr2   150-159 |          Gene1       intra</span></span>
<span><span class="co">#&gt;   [3]       Gene2 ---        chr2   250-259 |          Gene2       inter</span></span>
<span><span class="co">#&gt;   [4]       Gene3 ---        chr4   350-359 |          Gene3       intra</span></span>
<span><span class="co">#&gt;   [5]       Gene3 ---        chr4   450-459 |          Gene3       inter</span></span>
<span><span class="co">#&gt;        distance</span></span>
<span><span class="co">#&gt;       &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]        49</span></span>
<span><span class="co">#&gt;   [2]      &lt;NA&gt;</span></span>
<span><span class="co">#&gt;   [3]        50</span></span>
<span><span class="co">#&gt;   [4]      &lt;NA&gt;</span></span>
<span><span class="co">#&gt;   [5]        50</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 10 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 4 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ls_concat</span></span>
<span><span class="co">#&gt; linkSet object with 6 interactions and 2 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol      symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;   [2]      Gene41 ---        chr2   105-124 |          Gene2       Gene2</span></span>
<span><span class="co">#&gt;   [3]      Gene42 ---        chr3   205-224 |          Gene3       Gene3</span></span>
<span><span class="co">#&gt;   [4]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;   [5]      Gene41 ---        chr2   105-124 |          Gene2       Gene2</span></span>
<span><span class="co">#&gt;   [6]      Gene42 ---        chr3   205-224 |          Gene3       Gene3</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 6 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>The annotated distance and interaction type is shown in the
<code>distance</code> and <code>inter_type</code> column.</p>
<p>You can remove the intrachromosomal interactions by:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ls_concat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterLinks-linkSet-method.html">filterLinks</a></span><span class="op">(</span><span class="va">ls_concat</span>, filter_intra <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ls_concat</span></span>
<span><span class="co">#&gt; linkSet object with 2 interactions and 4 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | anchor1.symbol      symbol</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; |    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;   [2]      Gene40 ---        chr1      5-24 |          Gene1       Gene1</span></span>
<span><span class="co">#&gt;        inter_type  distance</span></span>
<span><span class="co">#&gt;       &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]       inter         9</span></span>
<span><span class="co">#&gt;   [2]       inter         9</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 2 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="annotations">Annotations<a class="anchor" aria-label="anchor" href="#annotations"></a>
</h2>
<p>It might be common that you only have the bait name, but don’t have
the exact bait region. The annotation function can help you to annotate
the bait region from a gene annotation.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bait</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"NONEXISTENT"</span><span class="op">)</span></span>
<span><span class="va">gr2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span><span class="op">)</span>,</span>
<span>                ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5000</span>, <span class="fl">6000</span>, <span class="fl">7000</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>,</span>
<span>                strand <span class="op">=</span> <span class="st">"+"</span><span class="op">)</span></span>
<span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/linkSet/man/linkSet-class.html" class="external-link">linkSet</a></span><span class="op">(</span><span class="va">bait</span>, <span class="va">gr2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Test annotatePromoter</span></span>
<span><span class="va">annotated_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/annotatePromoter-linkSet-method.html">annotatePromoter</a></span><span class="op">(</span><span class="va">ls</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, upstream <span class="op">=</span> <span class="fl">500</span>,overwrite <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="va">annotated_ls</span></span>
<span><span class="co">#&gt; linkSet object with 3 interactions and 0 metadata columns:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt;</span></span>
<span><span class="co">#&gt;   [1]       BRCA1 ---        chr1 5000-5099</span></span>
<span><span class="co">#&gt;   [2]        TP53 ---        chr2 6000-6099</span></span>
<span><span class="co">#&gt;   [3] NONEXISTENT ---        chr3 7000-7099</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 3 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="statistical-analysis">Statistical analysis<a class="anchor" aria-label="anchor" href="#statistical-analysis"></a>
</h2>
<p>Functional enhancers usually regulate multiple genes, we can use
cross gene analysis to identify the cross gene enhancers.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/crossGeneEnhancer-linkSet-method.html">crossGeneEnhancer</a></span><span class="op">(</span><span class="va">ls</span><span class="op">)</span></span>
<span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/orderLinks-linkSet-method.html">orderLinks</a></span><span class="op">(</span><span class="va">ls</span>,by <span class="op">=</span> <span class="st">"crossFreq"</span>,decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">ls</span></span>
<span><span class="co">#&gt; linkSet object with 3 interactions and 1 metadata column:</span></span>
<span><span class="co">#&gt;              bait     seqnames_oe ranges_oe | crossFreq</span></span>
<span><span class="co">#&gt;       &lt;character&gt;           &lt;Rle&gt; &lt;IRanges&gt; | &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]       BRCA1 ---        chr1 5000-5099 |         1</span></span>
<span><span class="co">#&gt;   [2]        TP53 ---        chr2 6000-6099 |         1</span></span>
<span><span class="co">#&gt;   [3] NONEXISTENT ---        chr3 7000-7099 |         1</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 3 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>With <code>plot_genomic_ranges</code> function, you can visualize the
link between bait and oe region.</p>
<p>You can choose the oe-centric views to visualize the functional
enhancer which regulate multiple genes.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_genomic_ranges-linkSet-method.html">plot_genomic_ranges</a></span><span class="op">(</span><span class="va">linkExample</span>,extend.base <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in plot_genomic_ranges(linkExample, extend.base = 10): score.col not</span></span>
<span><span class="co">#&gt; found, using count as default</span></span></code></pre></div>
<p><img src="linkSet_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>You can adjust the range of the views.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_genomic_ranges-linkSet-method.html">plot_genomic_ranges</a></span><span class="op">(</span><span class="va">linkExample</span>,extend.base <span class="op">=</span> <span class="fl">10</span>,x.range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>,<span class="fl">200</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in plot_genomic_ranges(linkExample, extend.base = 10, x.range = c(50, :</span></span>
<span><span class="co">#&gt; score.col not found, using count as default</span></span></code></pre></div>
<p><img src="linkSet_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gilbert Han.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
