<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>hic_workthrough • linkSet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="hic_workthrough">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">linkSet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/linkSet.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/hic_workthrough.html">hic_workthrough</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GilbertHan1011/linkSet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>hic_workthrough</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/GilbertHan1011/linkSet/blob/master/vignettes/hic_workthrough.Rmd" class="external-link"><code>vignettes/hic_workthrough.Rmd</code></a></small>
      <div class="d-none name"><code>hic_workthrough.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides a step-by-step guide to using the
<code>linkSet</code> package to analyze Hi-C/HiChIP data. We will use a
toy example to illustrate the main functions and workflows. Our goal is
to identify the enhancer-gene links in this example</p>
<p>We will use the following datasets as input:</p>
<ol style="list-style-type: decimal">
<li>validPairs produced by <a href="https://github.com/nservant/HiC-Pro" class="external-link">HiC-Pro</a>.</li>
<li>Mouse embryo body enhancer data from enhancer atlas <a href="http://www.enhanceratlas.org/" class="external-link">website</a>.</li>
<li>Gene annotation data from TxDb.Mmusculus.UCSC.mm10.knownGene
pakcage.</li>
</ol>
<p>We highly recommend you to use custom data instead of the example
data provided in this vignette.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GilbertHan1011/linkSet" class="external-link">linkSet</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Mmusculus.UCSC.mm10.knownGene</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Organism.dplyr</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>We use our custom function <code>readvalidPairs</code> to load the
example data. Firstly, we need to load into <a href="https://bioconductor.org/packages/release/bioc/vignettes/InteractionSet/inst/doc/GInteractions.html" class="external-link">GInteractions
object</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hic_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"toyValidatePair"</span>, </span>
<span>                        package<span class="op">=</span><span class="st">"linkSet"</span><span class="op">)</span></span>
<span><span class="va">gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readvalidPairs.html">readvalidPairs</a></span><span class="op">(</span><span class="va">hic_file</span><span class="op">)</span></span>
<span><span class="va">src</span> <span class="op">&lt;-</span> <span class="fu">Organism.dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Organism.dplyr/man/src.html" class="external-link">src_organism</a></span><span class="op">(</span><span class="st">"TxDb.Mmusculus.UCSC.mm10.knownGene"</span><span class="op">)</span></span>
<span><span class="co">#&gt; creating 'src_organism' database...</span></span>
<span><span class="va">geneGr</span> <span class="op">&lt;-</span> <span class="fu">Organism.dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">src</span>,columns <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="va">promoterGr</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">promoters</a></span><span class="op">(</span><span class="va">geneGr</span>,upstream <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 3 out-of-bound ranges located on sequences</span></span>
<span><span class="co">#&gt;   chr4_JH584294_random, chr4_JH584292_random, and chr5_GL456354_random.</span></span>
<span><span class="co">#&gt;   Note that ranges located on a sequence whose length is unknown (NA) or</span></span>
<span><span class="co">#&gt;   on a circular sequence are not considered out-of-bound (use</span></span>
<span><span class="co">#&gt;   seqlengths() and isCircular() to get the lengths and circularity flags</span></span>
<span><span class="co">#&gt;   of the underlying sequences). You can use trim() to trim these ranges.</span></span>
<span><span class="co">#&gt;   See ?`trim,GenomicRanges-method` for more information.</span></span>
<span><span class="va">file_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"http://www.enhanceratlas.org/data/download/enhancer/mm/Embryo_body.bed"</span><span class="op">)</span></span>
<span><span class="va">temp_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".bed"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="va">temp_file</span>, method <span class="op">=</span> <span class="st">"curl"</span><span class="op">)</span></span>
<span><span class="va">enhancer</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="va">temp_file</span><span class="op">)</span></span></code></pre></div>
<p>Because the hic data only contains digist end, so we resize the
region to upstream 5kb and downstream 5kb. After that, we use
<code>baitGInteractions</code> to generate the <code>linkSet</code>
object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/linkSet-GRange-Methods.html">resize</a></span><span class="op">(</span><span class="va">gi</span>,<span class="fl">10000</span>,fix <span class="op">=</span> <span class="st">"center"</span><span class="op">)</span></span>
<span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu">baitGInteractions</span><span class="op">(</span><span class="va">gi</span>,geneGr <span class="op">=</span> <span class="va">geneGr</span>,peakGr <span class="op">=</span> <span class="va">enhancer</span>,<span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="va">ls</span></span>
<span><span class="co">#&gt; linkSet object with 579 interactions and 2 metadata columns:</span></span>
<span><span class="co">#&gt;                bait     seqnames_oe         ranges_oe | anchor1.symbol</span></span>
<span><span class="co">#&gt;         &lt;character&gt;           &lt;Rle&gt;         &lt;IRanges&gt; |    &lt;character&gt;</span></span>
<span><span class="co">#&gt;     [1]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;     [2]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;     [3]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;     [4]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;     [5]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;     ...         ... ...         ...               ... .            ...</span></span>
<span><span class="co">#&gt;   [575]     Slco5a1 ---        chr1 14496641-14497000 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [576]     Slco5a1 ---        chr1 14496641-14497000 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [577]     Slco5a1 ---        chr1 14496641-14497000 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [578]     Slco5a1 ---        chr1 78187171-78187630 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [579]     Slco5a1 ---        chr1 39924881-39925280 |        Slco5a1</span></span>
<span><span class="co">#&gt;             anchor2.name</span></span>
<span><span class="co">#&gt;              &lt;character&gt;</span></span>
<span><span class="co">#&gt;     [1] 12.2266622787328</span></span>
<span><span class="co">#&gt;     [2] 12.2266622787328</span></span>
<span><span class="co">#&gt;     [3] 12.2266622787328</span></span>
<span><span class="co">#&gt;     [4] 12.2266622787328</span></span>
<span><span class="co">#&gt;     [5] 12.2266622787328</span></span>
<span><span class="co">#&gt;     ...              ...</span></span>
<span><span class="co">#&gt;   [575] 16.1891332694206</span></span>
<span><span class="co">#&gt;   [576] 16.1891332694206</span></span>
<span><span class="co">#&gt;   [577] 16.1891332694206</span></span>
<span><span class="co">#&gt;   [578]  11.771392591681</span></span>
<span><span class="co">#&gt;   [579] 14.6843556168742</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 109 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 66 sequences (1 circular) from mm10 genome</span></span></code></pre></div>
<p>When we print the <code>linkSet</code> object, we can see the basic
information of the <code>linkSet</code> object. By default, we don’t
show the bait region. But you are interested in the bait region, you can
set <code>showBaitRegion = TRUE</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">showLinkSet</span><span class="op">(</span><span class="va">ls</span>, baitRegion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; linkSet object with 579 interactions and 2 metadata columns:</span></span>
<span><span class="co">#&gt;          bait seqnames_bait       ranges_bait     seqnames_oe         ranges_oe</span></span>
<span><span class="co">#&gt;   [1]   Sulf1          chr1 12692277-12861192 ---        chr1 12785091-12785750</span></span>
<span><span class="co">#&gt;   [2]   Sulf1          chr1 12692277-12861192 ---        chr1 12785091-12785750</span></span>
<span><span class="co">#&gt;   [3]   Sulf1          chr1 12692277-12861192 ---        chr1 12785091-12785750</span></span>
<span><span class="co">#&gt;   [4]   Sulf1          chr1 12692277-12861192 ---        chr1 12785091-12785750</span></span>
<span><span class="co">#&gt;   [5]   Sulf1          chr1 12692277-12861192 ---        chr1 12785091-12785750</span></span>
<span><span class="co">#&gt;   ...     ...           ...               ... ...         ...               ...</span></span>
<span><span class="co">#&gt; [575] Slco5a1          chr1 12866549-12992650 ---        chr1 14496641-14497000</span></span>
<span><span class="co">#&gt; [576] Slco5a1          chr1 12866549-12992650 ---        chr1 14496641-14497000</span></span>
<span><span class="co">#&gt; [577] Slco5a1          chr1 12866549-12992650 ---        chr1 14496641-14497000</span></span>
<span><span class="co">#&gt; [578] Slco5a1          chr1 12866549-12992650 ---        chr1 78187171-78187630</span></span>
<span><span class="co">#&gt; [579] Slco5a1          chr1 12866549-12992650 ---        chr1 39924881-39925280</span></span>
<span><span class="co">#&gt;       | anchor1.symbol     anchor2.name</span></span>
<span><span class="co">#&gt;   [1] |          Sulf1 12.2266622787328</span></span>
<span><span class="co">#&gt;   [2] |          Sulf1 12.2266622787328</span></span>
<span><span class="co">#&gt;   [3] |          Sulf1 12.2266622787328</span></span>
<span><span class="co">#&gt;   [4] |          Sulf1 12.2266622787328</span></span>
<span><span class="co">#&gt;   [5] |          Sulf1 12.2266622787328</span></span>
<span><span class="co">#&gt;   ... .            ...              ...</span></span>
<span><span class="co">#&gt; [575] |        Slco5a1 16.1891332694206</span></span>
<span><span class="co">#&gt; [576] |        Slco5a1 16.1891332694206</span></span>
<span><span class="co">#&gt; [577] |        Slco5a1 16.1891332694206</span></span>
<span><span class="co">#&gt; [578] |        Slco5a1  11.771392591681</span></span>
<span><span class="co">#&gt; [579] |        Slco5a1 14.6843556168742</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="diagnose-and-filter-links">Diagnose and filter links<a class="anchor" aria-label="anchor" href="#diagnose-and-filter-links"></a>
</h2>
<p>Now, we can run diagnoseLinkSet to check the distance distribution
and inter/intra interaction percentage.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">diagnoseLinkSet</span><span class="op">(</span><span class="va">ls</span><span class="op">)</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre></div>
<p><img src="hic_workthrough_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; linkSet object with 579 interactions and 4 metadata columns:</span></span>
<span><span class="co">#&gt;                bait     seqnames_oe         ranges_oe | anchor1.symbol</span></span>
<span><span class="co">#&gt;         &lt;character&gt;           &lt;Rle&gt;         &lt;IRanges&gt; |    &lt;character&gt;</span></span>
<span><span class="co">#&gt;     [1]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;     [2]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;     [3]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;     [4]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;     [5]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;     ...         ... ...         ...               ... .            ...</span></span>
<span><span class="co">#&gt;   [575]     Slco5a1 ---        chr1 14496641-14497000 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [576]     Slco5a1 ---        chr1 14496641-14497000 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [577]     Slco5a1 ---        chr1 14496641-14497000 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [578]     Slco5a1 ---        chr1 78187171-78187630 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [579]     Slco5a1 ---        chr1 39924881-39925280 |        Slco5a1</span></span>
<span><span class="co">#&gt;             anchor2.name  inter_type  distance</span></span>
<span><span class="co">#&gt;              &lt;character&gt; &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;     [1] 12.2266622787328       inter      8686</span></span>
<span><span class="co">#&gt;     [2] 12.2266622787328       inter      8686</span></span>
<span><span class="co">#&gt;     [3] 12.2266622787328       inter      8686</span></span>
<span><span class="co">#&gt;     [4] 12.2266622787328       inter      8686</span></span>
<span><span class="co">#&gt;     [5] 12.2266622787328       inter      8686</span></span>
<span><span class="co">#&gt;     ...              ...         ...       ...</span></span>
<span><span class="co">#&gt;   [575] 16.1891332694206       inter   1567221</span></span>
<span><span class="co">#&gt;   [576] 16.1891332694206       inter   1567221</span></span>
<span><span class="co">#&gt;   [577] 16.1891332694206       inter   1567221</span></span>
<span><span class="co">#&gt;   [578]  11.771392591681       inter  65257801</span></span>
<span><span class="co">#&gt;   [579] 14.6843556168742       inter  26995481</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 109 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 66 sequences (1 circular) from mm10 genome</span></span></code></pre>
<p>Intrachromosomal interaction and long distance interaction are likely
be noise, so we filter them.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterLinks-linkSet-method.html">filterLinks</a></span><span class="op">(</span><span class="va">ls</span>,filter_intra <span class="op">=</span> <span class="cn">TRUE</span>, filter_unannotate <span class="op">=</span> <span class="cn">TRUE</span>, distance <span class="op">=</span> <span class="fl">50000000</span><span class="op">)</span></span></code></pre></div>
<p>Duplicated links are associated with contact frequency, so it’s a
good idea to count duplicated links.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu">countInteractions</span><span class="op">(</span><span class="va">ls</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/orderLinks-linkSet-method.html">orderLinks</a></span><span class="op">(</span><span class="va">ls</span>,by <span class="op">=</span> <span class="st">"count"</span>,decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt; linkSet object with 50 interactions and 5 metadata columns:</span></span>
<span><span class="co">#&gt;               bait     seqnames_oe         ranges_oe | anchor1.symbol</span></span>
<span><span class="co">#&gt;        &lt;character&gt;           &lt;Rle&gt;         &lt;IRanges&gt; |    &lt;character&gt;</span></span>
<span><span class="co">#&gt;    [1]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;    [2]     Slco5a1 ---        chr1 12785091-12785750 |        Slco5a1</span></span>
<span><span class="co">#&gt;    [3]     Slco5a1 ---        chr1 14496641-14497000 |        Slco5a1</span></span>
<span><span class="co">#&gt;    [4]       Ncoa2 ---        chr1 12785091-12785750 |          Ncoa2</span></span>
<span><span class="co">#&gt;    [5]       Sulf1 ---        chr1 14496641-14497000 |          Sulf1</span></span>
<span><span class="co">#&gt;    ...         ... ...         ...               ... .            ...</span></span>
<span><span class="co">#&gt;   [46]     Slco5a1 ---        chr1 35926611-35926860 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [47]     Slco5a1 ---        chr1 38683021-38683930 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [48]     Slco5a1 ---        chr1 34163531-34164100 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [49]     Slco5a1 ---        chr1 62841301-62841760 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [50]     Slco5a1 ---        chr1 39924881-39925280 |        Slco5a1</span></span>
<span><span class="co">#&gt;            anchor2.name  inter_type  distance     count</span></span>
<span><span class="co">#&gt;             &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;    [1] 12.2266622787328       inter      8686       404</span></span>
<span><span class="co">#&gt;    [2] 12.2266622787328       inter    144179        20</span></span>
<span><span class="co">#&gt;    [3] 16.1891332694206       inter   1567221        10</span></span>
<span><span class="co">#&gt;    [4] 12.2266622787328       inter    471173         8</span></span>
<span><span class="co">#&gt;    [5] 16.1891332694206       inter   1720086         5</span></span>
<span><span class="co">#&gt;    ...              ...         ...       ...       ...</span></span>
<span><span class="co">#&gt;   [46] 14.3114832520841       inter  22997136         1</span></span>
<span><span class="co">#&gt;   [47] 10.9529360911238       inter  25753876         1</span></span>
<span><span class="co">#&gt;   [48] 11.2903880454209       inter  21234216         1</span></span>
<span><span class="co">#&gt;   [49] 11.6972535378061       inter  49911931         1</span></span>
<span><span class="co">#&gt;   [50] 14.6843556168742       inter  26995481         1</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 48 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 66 sequences (1 circular) from mm10 genome</span></span></code></pre></div>
<p>We can notice that there is a significant link strength between
<code>Sulf1</code> and <code>chr1:12785091-12785750</code>.</p>
</div>
<div class="section level2">
<h2 id="cross-gene-links-and-visualization">Cross gene links and visualization<a class="anchor" aria-label="anchor" href="#cross-gene-links-and-visualization"></a>
</h2>
<p>Enhancers that regulate multiple genes are biologically
meaningful.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/crossGeneEnhancer-linkSet-method.html">crossGeneEnhancer</a></span><span class="op">(</span><span class="va">ls</span><span class="op">)</span></span>
<span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/orderLinks-linkSet-method.html">orderLinks</a></span><span class="op">(</span><span class="va">ls</span>,by <span class="op">=</span> <span class="st">"crossFreq"</span>,decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">ls</span></span>
<span><span class="co">#&gt; linkSet object with 50 interactions and 6 metadata columns:</span></span>
<span><span class="co">#&gt;               bait     seqnames_oe         ranges_oe | anchor1.symbol</span></span>
<span><span class="co">#&gt;        &lt;character&gt;           &lt;Rle&gt;         &lt;IRanges&gt; |    &lt;character&gt;</span></span>
<span><span class="co">#&gt;    [1]       Sulf1 ---        chr1 12785091-12785750 |          Sulf1</span></span>
<span><span class="co">#&gt;    [2]     Slco5a1 ---        chr1 12785091-12785750 |        Slco5a1</span></span>
<span><span class="co">#&gt;    [3]        Xkr9 ---        chr1 12785091-12785750 |           Xkr9</span></span>
<span><span class="co">#&gt;    [4]      Ogfrl1 ---        chr1 12785091-12785750 |         Ogfrl1</span></span>
<span><span class="co">#&gt;    [5]      Adgrb3 ---        chr1 12785091-12785750 |         Adgrb3</span></span>
<span><span class="co">#&gt;    ...         ... ...         ...               ... .            ...</span></span>
<span><span class="co">#&gt;   [46]     Slco5a1 ---        chr1 39758181-39758690 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [47]     Slco5a1 ---        chr1 38683021-38683930 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [48]     Slco5a1 ---        chr1 34163531-34164100 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [49]     Slco5a1 ---        chr1 62841301-62841760 |        Slco5a1</span></span>
<span><span class="co">#&gt;   [50]     Slco5a1 ---        chr1 39924881-39925280 |        Slco5a1</span></span>
<span><span class="co">#&gt;            anchor2.name  inter_type  distance     count crossFreq</span></span>
<span><span class="co">#&gt;             &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;    [1] 12.2266622787328       inter      8686       404        37</span></span>
<span><span class="co">#&gt;    [2] 12.2266622787328       inter    144179        20        37</span></span>
<span><span class="co">#&gt;    [3] 12.2266622787328       inter    899826         2        37</span></span>
<span><span class="co">#&gt;    [4] 12.2266622787328       inter  10596562         1        37</span></span>
<span><span class="co">#&gt;    [5] 12.2266622787328       inter  12663171         5        37</span></span>
<span><span class="co">#&gt;    ...              ...         ...       ...       ...       ...</span></span>
<span><span class="co">#&gt;   [46]  11.037989817782       inter  26828836         1         1</span></span>
<span><span class="co">#&gt;   [47] 10.9529360911238       inter  25753876         1         1</span></span>
<span><span class="co">#&gt;   [48] 11.2903880454209       inter  21234216         1         1</span></span>
<span><span class="co">#&gt;   [49] 11.6972535378061       inter  49911931         1         1</span></span>
<span><span class="co">#&gt;   [50] 14.6843556168742       inter  26995481         1         1</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 48 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 66 sequences (1 circular) from mm10 genome</span></span></code></pre></div>
<p>We can use <code>plot_genomic_ranges</code> to visualize the cross
gene links.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_genomic_ranges-linkSet-method.html">plot_genomic_ranges</a></span><span class="op">(</span><span class="va">ls</span>,showOE <span class="op">=</span> <span class="fu">oe</span><span class="op">(</span><span class="va">ls</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="hic_workthrough_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>We can also choose to visualze the bait center region.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_genomic_ranges-linkSet-method.html">plot_genomic_ranges</a></span><span class="op">(</span><span class="va">ls</span>,showBait <span class="op">=</span> <span class="st">"Ncoa2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="hic_workthrough_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_genomic_ranges-linkSet-method.html">plot_genomic_ranges</a></span><span class="op">(</span><span class="va">ls</span>,showBait <span class="op">=</span> <span class="st">"Sulf1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="hic_workthrough_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gilbert Han.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
